### ISSUE TITLE
RouteReuseStrategy - retrieve fired before shouldAttach

### ISSUE THREAD

**Issue by kaanoo2904** [angular/angular#18622]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior

1. navigate to a page in app.
2.  `retrieve` in `CustomReuseStrategy` fired before `shouldAttach` check.

![screenshot from 2017-08-10 11-28-40](https://user-images.githubusercontent.com/10462294/29154502-3287cf1e-7dbf-11e7-8fde-6b1fecffca60.png)

I've test with my own project ( which use Angular 4), or with a demo project from this article (with Angular 2): https://www.softwarearchitekt.at/post/2016/12/02/sticky-routes-in-angular-2-3-with-routereusestrategy.aspx
(this demo project can be found [here](https://github.com/manfredsteyer/angular-2-reuse-strategy-sample))
## Expected behavior

Should `retrieve` fired only after `shouldAttach` return true ?

## Environment

<pre><code>
Angular version: 
- demo project: 2.3.0
- my own project: 4.0.0

Browser:
- [ ] Chrome (desktop) version 60

For Tooling issues:
- Node version: 8.1.2  
- Platform:  Ubuntu 16.04

Others:

</code></pre>

**Comment by ofriedrich** [comment]
`shouldReuseRoute` is called before the router navigates to its new target. It asks if a snapshot from the current route could be reused for the future route. Afterwards the router navigates to the new route. So `shouldAttach`, which should return true, if any of your stored snapshots matches, is called after `shouldReuseRoute `.

For me it looks like everything is fine.

**Comment by kaanoo2904** [comment]
@ofriedrich  Thanks for reply :) but my post is about `retrieve` fired before `shouldAttach`. Is it a normal behavior ?

**Comment by jasonaden** [comment]
It appears this behaves as expected. If you still feel there is an issue, please provide further details in a new issue.

<hr>

Hello, we reviewed this issue and determined that it doesn't fall into the bug report or feature request category. This issue tracker is not suitable for support requests, please repost your issue on [StackOverflow](http://stackoverflow.com/) using tag `angular`.

If you are wondering why we don't resolve support issues via the issue tracker, please [check out this explanation](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#-got-a-question-or-problem).

**Comment by StevenLiekens** [comment]
@jasonaden there seems to be a lot of misunderstanding in this thread so can you clarify some things?

I implemented my own RouteReuseStrategy and I also believe that something isn't right about the order of events.

This is the order of events that I recall.

When creating a new router state:
1. router calls `shouldReuseRoute`
    * if return value is true: reuse the current route (and skip other steps)
    * if return value is false: goto step 2.
2. router calls `retrieve`
    * if return value is null: build a new snapshot (and skip other steps)
    * if return value is a detached route handle: goto step 3
3. router calls `shouldAttach`
    * if return value is false: ??? (never tested this but presumably router destroys the snapshot?)
    * if return value is true: goto step 4
4. router calls `retrieve` (again)
    * if return value is null: ??? (never tested this but presumably results in a crash?)
    * if return value is a detached route handle: reattach it

Does this sound more or less right to you? I typed all of this from memory so I may have made a mistake or two.

**Comment by albanx** [comment]
We definitely need some documentation and clarification on this provider. It is not clear how to use RouteReuseStrategy, angular.io provides no examples

**Comment by iamjamesjose** [comment]
> @jasonaden there seems to be a lot of misunderstanding in this thread so can you clarify some things?
> 
> I implemented my own RouteReuseStrategy and I also believe that something isn't right about the order of events.
> 
> This is the order of events that I recall.
> 
> When creating a new router state:
> 
> 1. router calls `shouldReuseRoute`
>    
>    * if return value is true: reuse the current route (and skip other steps)
>    * if return value is false: goto step 2.
> 2. router calls `retrieve`
>    
>    * if return value is null: build a new snapshot (and skip other steps)
>    * if return value is a detached route handle: goto step 3
> 3. router calls `shouldAttach`
>    
>    * if return value is false: ??? (never tested this but presumably router destroys the snapshot?)
>    * if return value is true: goto step 4
> 4. router calls `retrieve` (again)
>    
>    * if return value is null: ??? (never tested this but presumably results in a crash?)
>    * if return value is a detached route handle: reattach it
> 
> Does this sound more or less right to you? I typed all of this from memory so I may have made a mistake or two.

Hi Sir,
Did you get any solution to this problem? I am also facing some issue on implementing custom 

This is my StackOverflow question.
https://stackoverflow.com/questions/57490593/angular-custom-routereusestrategy-retrieve-method-fails-with-error-cannot-reatt

Thanks in advance.
